cmd/version.go

```diff
@@ -16,11 +16,64 @@ package cmd
 
 //go:generate ./mkversion.sh
 
-// Version will be overwritten at build-time via mkversion.sh
-var Version = "unknown"
+import (
+	"io/ioutil"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+)
 
-func MockVersion(version string) (restore func()) {
-	old := Version
-	Version = version
-	return func() { Version = old }
+var (
+	// Version will be overwritten at build-time via mkversion.sh
+	Version = "unknown"
+
+	pid2ProcPath = "/proc/2/status"
+	selfPid      = os.Getpid()
+
+	containerOnce    sync.Once
+	containerRuntime bool = true
+)
+
+// Containerised returns true if we are running inside a container runtime
+// such as lxd or Docker. The detection is only intended for Linux host
+// environments as it looks for the Linux kernel kthreadd process, started
+// by the Linux kernel after init. Kernel initialised processes do not have
+// a parent as its not a child of init (PID1) and as a result has its ppid
+// set to zero. We use this property to detect the absence of a container
+// runtime. If /proc is not mounted, the detection will assume its a
+// container, so make sure early mounts are completed before running this
+// check on an unconfined host.
+func Containerised() bool {
+	containerOnce.Do(func() {
+		if s, err := ioutil.ReadFile(pid2ProcPath); err == nil {
```

Comment:
I am not 100% sure this would always work. Here's an example:

```bash
$ docker run --rm --pid host ubuntu sh -c "[ -d /proc/2 ] && ps -AfwT" | head -3
UID          PID    SPID    PPID  C STIME TTY          TIME CMD
root           1       1       0  0 Jun23 ?        00:09:13 /sbin/init
root           2       2       0  0 Jun23 ?        00:00:00 [kthreadd]
```

Here are some other reasons that make me think this might not be a universal rule:
 - the fact that containers have a PID 1 and then other processes with PID >2 is due to the container runtime's way of managing the container. Afaiu, this implementation may differ from runtime to runtime, so, can we be absolutely sure that we won't ever have a PID 2 inside a container (today or in the future)?
 - some runtimes allow the deployment-time injection of an init process into the containers. I've tried this with Docker (`docker run --init --rm -it ubuntu`), and PID 2 was still not present...but could it be though? Could the introduction of init systems into the container, by the container runtime, void this assessment?

Reply:
@cjdcordeiro Thanks for this! Let me look into this. What did you do to be able to see the real PID2 inside the container ?

Reply:
We look for `/.dockerenv` [1] and `/run/.containerenv` [2] instead of or in addition to checking for kernel threads. The latter is apparently not stable interface and may disappear but it's still being created and scripts depend on it, e.g. remediation scripts generated by OpenSCAP.

[1] https://superuser.com/questions/1021834/what-are-dockerenv-and-dockerinit
[2] https://docs.podman.io/en/latest/markdown/podman-run.1.html

Reply:
I have a feeling the proper answer is no single solution is bulletproof. 

http://git.annexia.org/?p=virt-what.git;a=blob;f=virt-what.in;h=896bfd8b3cb40e8c265597c33eec5494e6fcfd49;hb=HEAD

@cjdcordeiro @woky I wonder if we should start by covering a list of cases from here, and then keep it up to date as technologies change. Anything we use is bound to stop working at some point, so I am thinking we should have a list of checks to detect containers, otherwise we assume its hardware/vm?

Reply:
(above is the red-hat virt-what source)

I will respin this PR with the latest feedback, breaking out this container detection into its own file, and structure it so that checks can easily be added.

Reply:
> Anything we use is bound to stop working at some point, so I am thinking we should have a list of checks to detect containers, otherwise we assume its hardware/vm?

Sounds good. The isolation is too good to be detected reliably. :-)

Reply:
New proposal to follow soon. Resolving this for now.